Add GetFeatGrantLevel, GetInternalObjectType, PluginExists, GetUserDirectory

Jorteck opened the pull request:
**Creature - GetFeatGrantLevel**
When a feat is removed from a creature, there is no easy API to determine which level this feat was removed from, if we need to restore it later using AddFeatByLevel.

**Object - GetInternalObjectType**
Modules, Areas and a few other object types do not have their own OBJECT_TYPE_* constant that is returned from GetObjectType. NWNX knows this exact type, and this would be super handy to have for type safety in a managed implementation (e.g. DotNet).

**Util - PluginExists**
Currently there is no way to check if a plugin exists (e.g. to check if a plugin is an optional dependency) without throwing an error. I'm not sure if this might make more sense as a core function.

**Util - GetUserDirectory**
When using other plugins for databases and persistent data (e.g. EF in DotNet), this can be useful for getting a location to store such data.

Daztek commented on Plugins/Util/NWScript/nwnx_util.nss:
```diff
@@ -163,6 +163,15 @@ int NWNX_Util_RegisterServerConsoleCommand(string sCommand, string sScriptChunk)
 void NWNX_Util_UnregisterServerConsoleCommand(string sCommand);
 
+/// @brief Determines if the given plugin exists and is enabled.
+/// @param sPlugin The name of the plugin to check.
```
Should probably note that it expects the plugin name as `NWNX_{Name}`, case-sensitive and all.

Unless you want to go the extra mile and strip the prefix only when needed and use `GetServices()->m_plugins->GetCanonicalPluginName()` to deal with case-sensitivity. :P

Jorteck replied:
I will add a comment ;)

Removing the ` NWNX_*` prefix might be a bit much, I think it makes sense to match the plugin naming used in `NWNX_Core`.

mtijanic commented on Plugins/Creature/Creature.cpp:
```diff
@@ -266,6 +267,38 @@ ArgumentStack Creature::GetFeatByLevel(ArgumentStack&& args)
+            {
+                if (pLevelStats->m_lstFeats.element[j] != uFeat)
+                {
+                    continue;
```
Style nit: rather than `if (x != y) continue;`, it's clearer when you do `if (x == y) { do_thing(); }`.

mtijanic commented on Plugins/Creature/Creature.cpp:
```diff
@@ -266,6 +267,38 @@ ArgumentStack Creature::GetFeatByLevel(ArgumentStack&& args)
+                }
+
+                Services::Events::InsertArgument(stack, i + 1);
+                return stack;
```
Style nit: rest of this file uses oneline `return Services::Events::Arguments(i+1);`

mtijanic commented on Plugins/Object/NWScript/nwnx_object.nss:
```diff
@@ -299,6 +321,11 @@ void NWNX_Object_DeleteVarRegex(object oObject, string sRegex);
+/// @brief Gets the given object's internal type (NWNX_OBJECT_TYPE_INTERNAL_*)
+/// @param oObject The object.
+/// @return The object's type (NWNX_OBJECT_TYPE_INTERNAL_*)
+int NWNX_Object_GetInternalObjectType(object oObject);
```
See https://github.com/nwnxee/unified/blob/master/Plugins/Tweaks/Tweaks/StringToIntBaseToAuto.cpp

This is prior art for just modifying existing nwscript functions to do something else. I'd rather just extend the default `GetObjectType()` to return `OBJECT_TYPE_AREA` and `OBJECT_TYPE_MODULE` in addition to the other stuff.

Also, this is already possible to do in pure nwscript, by just checking something like:
```C
int myGetObjectType(object oObj) {
  if (oObj == GetModule())
    return OBJECT_TYPE_MODULE;
  if (GetIsAreaAboveGround(oObj) != AREA_INVALID);
    return OBJECT_TYPE_AREA;
  return GetObjectType(oObj);
}
```

mtijanic commented on Plugins/Object/Object.cpp:
```diff
@@ -839,4 +840,22 @@ ArgumentStack Object::GetPositionIsInTrigger(ArgumentStack&& args)
+    const auto objectId = Services::Events::ExtractArgument<Types::ObjectID>(args);
+    ArgumentStack stack;
+
+    if (objectId == Constants::OBJECT_INVALID)
```
style: invert logic.
```C
{
    int retVal = -1;
    const auto objectId = ...;
    if (auto* go = Utils::GetGameObject(objectId))
    {
        retVal = go->m_nObjectType;
    }
    return Services::Events::Arguments(retVal);
}

mtijanic commented on Plugins/Util/Util.cpp:
```diff
@@ -548,4 +552,31 @@ ArgumentStack Util::UnregisterServerConsoleCommand(ArgumentStack&& args)
+
+    if (GetServices()->m_plugins->FindPluginByName(pluginNameWithoutPrefix))
+    {
+        Services::Events::InsertArgument(stack, 1);
```
Style: As above, you can invert the logic to simplify it.

mtijanic commented on Plugins/Util/Util.cpp:
```diff
@@ -548,4 +552,31 @@ ArgumentStack Util::UnregisterServerConsoleCommand(ArgumentStack&& args)
+ArgumentStack Util::GetUserDirectory(ArgumentStack&&)
+{
+    ArgumentStack stack;
+    char* retval = Globals::ExoBase()->m_sUserDirectory.CStr();
```
Style: Oneliner:
```C
return Services::Events::Arguments(Globals::ExoBase()->m_sUserDirectory.CStr());
```

mtijanic commented on Plugins/Util/NWScript/nwnx_util.nss:
```diff
@@ -163,6 +163,15 @@ int NWNX_Util_RegisterServerConsoleCommand(string sCommand, string sScriptChunk)
+/// @brief Determines if the given plugin exists and is enabled.
+/// @param sPlugin The name of the plugin to check.
+/// @return TRUE if the plugin exists and is enabled, otherwise FALSE.
+int NWNX_Util_PluginExists(string sPlugin);
```
I'm not thrilled that this is part of util, but probably good enough for now.

Jorteck commented:
I've fixed up the points above and squashed those commits :)

> See https://github.com/nwnxee/unified/blob/master/Plugins/Tweaks/Tweaks/StringToIntBaseToAuto.cpp
> 
> This is prior art for just modifying existing nwscript functions to do something else. I'd rather just extend the default `GetObjectType()` to return `OBJECT_TYPE_AREA` and `OBJECT_TYPE_MODULE` in addition to the other stuff.
> 
> Also, this is already possible to do in pure nwscript, by just checking something like:
> 
> ```c
> int myGetObjectType(object oObj) {
>   if (oObj == GetModule())
>     return OBJECT_TYPE_MODULE;
>   if (GetIsAreaAboveGround(oObj) != AREA_INVALID);
>     return OBJECT_TYPE_AREA;
>   return GetObjectType(oObj);
> }
> ```

What would be the nicest way to add these other object types to GetObjectType without breaking existing scripts? I'm also not sure if there's a way to determine if something is a valid sound object for `SoundObjectPlay()`

mtijanic commented:
Merging now, and we can discuss about extending GetObjectType later. Thanks!
